server/src/test/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorServiceTests.java

```diff
@@ -0,0 +1,301 @@
+                )
+            ),
+            List.of()
+        );
+        var service = createAllocationHealthIndicatorService(clusterState);
+
+        assertThat(
+            service.calculate(),
+            equalTo(
+                createExpectedResult(
+                    YELLOW,
+                    unavailableReplicas.size() > 1
+                        ? "This cluster has " + unavailableReplicas.size() + " unavailable replicas."
+                        : "This cluster has 1 unavailable replica.",
+                    Map.of(
+                        "started_primaries",
+                        1,
+                        "unassigned_replicas",
+                        unavailableReplicas.size(),
+                        "started_replicas",
+                        availableReplicas.size()
+                    )
+                )
+            )
+        );
+    }
+
+    public void testShouldBeRedWhenThereAreUnassignedPrimaries() {
+        var clusterState = createClusterStateWith(List.of(index("red-index", new ShardAllocation(randomNodeId(), UNAVAILABLE))), List.of());
+        var service = createAllocationHealthIndicatorService(clusterState);
+
+        assertThat(
+            service.calculate(),
+            equalTo(createExpectedResult(RED, "This cluster has 1 unavailable primary.", Map.of("unassigned_primaries", 1)))
+        );
+    }
+
+    public void testShouldBeGreenWhenThereAreRestartingReplicas() {
+        var clusterState = createClusterStateWith(
+            List.of(
+                index(
+                    "restarting-index",
+                    new ShardAllocation(randomNodeId(), AVAILABLE),
+                    new ShardAllocation("node-0", RESTARTING, System.currentTimeMillis() - 1_000)
+                )
+            ),
+            List.of(new NodeShutdown("node-0", RESTART, 60))
+        );
+        var service = createAllocationHealthIndicatorService(clusterState);
+
+        assertThat(
+            service.calculate(),
+            equalTo(
+                createExpectedResult(
+                    GREEN,
+                    "This cluster has 1 restarting replica.",
+                    Map.of("started_primaries", 1, "restarting_replicas", 1)
+                )
+            )
+        );
+    }
+
+    public void testShouldBeGreenWhenThereAreRestartingPrimaries() {
+        var clusterState = createClusterStateWith(
+            List.of(index("restarting-index", new ShardAllocation("node-0", RESTARTING, System.currentTimeMillis() - 1_000))),
+            List.of(new NodeShutdown("node-0", RESTART, 60))
+        );
+        var service = createAllocationHealthIndicatorService(clusterState);
+
+        assertThat(
+            service.calculate(),
+            equalTo(createExpectedResult(GREEN, "This cluster has 1 restarting primary.", Map.of("restarting_primaries", 1)))
+        );
+    }
+
+    public void testShouldBeRedWhenRestartingPrimariesReachedAllocationDelay() {
```

Comment:
I think we should add this for replicas too (checking for yellow status when restart has expired).